### ISSUE TITLE
Review uses of ctype macros inside common_opencl

### ISSUE THREAD

**Issue by claudioandre-br** [openwall/john#3659]
From #3657. I found `isdigit` and `isspace` use cases.
********

It's not allowed to use ctype macros on "char" - they're to be used on "int" - and Intel isn't to blame), and we should have this code depend on current locale.

```
while (isspace(ARCH_INDEX(*p))) /* Intel quirk */
 	p++;
```
So this code is ideally changed to:
```
while (*p == ' ' || *p == '\t')
        p++;
```

See also #1847

**Comment by solardiz** [comment]
Of course, I meant we "shouldn't" depend on current locale. (I edited my original comment now.)

If we do use ctype macros (want to depend on current locale somewhere?), then a proper way to do so would be e.g. `isalpha((int)(unsigned char)*p)`. But I think in JtR it's best not to depend on locale at all.

**Comment by jfoug** [comment]
I would think the best way (in JtR) to do these type isalpha, is to revert back to the very old implementation of the is*() macros (prior to locale.h).  In that mode, we simply have an array of 256 unsigned. These unsigned are simply set with bits.  This array should be pre-built (so it is always properly set).  Then the macros are simply something like this:

```c
// jtr_ctype.h
#define _IS_DIGIT 1
#define _IS_LOWER 2
#define _IS_UPPER 4
#define _IS_ALPHA (_IS_LOWER|_IS_UPPER)
#define _IS_HEXU 8
#define _IS_HEXL 0x10
#define _IS_HEX (_IS_HEXU|_IS_HEXL)
#define _IS_CTRL 0x20
// etc

extern unsigned _IS_TYPES[256];

#define jtr_isdigit(x)  ( _IS_TYPES[(unsigned char)(unsigned)(x)] & _IS_DIGIT)
#define jtr_islower(x) ( _IS_TYPES[(unsigned char)(unsigned)(x)] & _IS_LOWER)
#define jtr_isalpha(x) ( _IS_TYPES[(unsigned char)(unsigned)(x)] & _IS_ALPHA)
// ... etc


// jtr_ctype.c  (note, data only, no code. All code in macros)
#include "jtr_ctype.h"

// this array was pre-built, based on ASCII values.
unsigned _IS_TYPES[256] = { _IS_CTRL, _IS_CTRL, _IS_CTRL, ....... };

```

If done in this very old C fashion, the code is VERY fast, does not depend in the least bit on any locale (which can make things change significantly, can be used with char (7-bit only), with unsigned char, int/unsigned (if value is restricted from 0 to 256).

This is the circa 1980's ctype.h method anyway, lol, before anyone gave a crap about internationalization.   This is VERY similar to how we do are is*() common functions (which could easily be done with a single array and set bits).

Also, if done this way, it makes retrofitting code from someone else, to use these locale-unfriendly function UNLESS the locale methods are actually needed.

**Comment by solardiz** [comment]
> If we do use ctype macros (want to depend on current locale somewhere?), then a proper way to do so would be e.g. `isalpha((int)(unsigned char)*p)`.

I've implemented this across the entire tree in ee5efb483c7674b080afae15ed4980d2d624a2ed, which I think is good enough to close this issue.

> But I think in JtR it's best not to depend on locale at all.

We're inconsistent about this now, but we reset to the `C` locale as a workaround, in `john.c`:

```c
#if (!AC_BUILT || HAVE_LOCALE_H)
        if (setlocale(LC_CTYPE, "")) {
                char *parsed = setlocale(LC_CTYPE, NULL);

                if (parsed) {
                        char *p;

                        john_terminal_locale = str_alloc_copy(parsed);
                        if ((p = strchr(john_terminal_locale, '.')))
                                parsed = ++p;
                        if (strcmp(parsed, "C"))
                                options.terminal_enc = cp_name2id(parsed, 0);
                }
#if HAVE_OPENCL
                if (options.terminal_enc)
                        sprintf(gpu_degree_sign, "%ls", DEGREE_SIGN);
#endif
                /* We misuse ctype macros so this must be reset */
                setlocale(LC_CTYPE, "C");
        }
#endif
```

except that in the deprecated(?), not built by default `rexgen` mode we do:

```c
static void rexgen_setlocale()
{
        char *john_locale;
        const char *ret;

        if (options.internal_cp == UTF_8)
                john_locale = "en_US.UTF-8";
        else
                john_locale = "C";

        ret = setlocale(LC_CTYPE, john_locale);

        if (options.verbosity >= VERB_MAX) {
                if (ret)
                        fprintf(stderr, "regex: Locale set to %s\n", ret);
                else
                        fprintf(stderr, "regex: Failed to set locale \"%s\"\n", john_locale);
        }
}
```
